[WGSL] Add overloads for vecN constructors

tadeuzagallo opened the pull request:
#### 4b417b85f4bf9a1f51f0341bf87a5fce436907d5
<pre>
[WGSL] Add overloads for vecN constructors
<a href="https://bugs.webkit.org/show_bug.cgi?id=254586">https://bugs.webkit.org/show_bug.cgi?id=254586</a>
rdar://107315663

Reviewed by Myles C. Maxfield.

Add the declarations for all the overloads of vec2, vec3 and vec4 constructors.
This requires introducing a couple new constraints, although constraints aren&apos;t
validated yet. There&apos;s only one overload per constructor disabled with a fixme,
since they require explicitly specifying the type (e.g. vecN&lt;T&gt;(...)), but we
don&apos;t currently support that, but will be implemented next.

* Source/WebGPU/WGSL/Overload.h:
* Source/WebGPU/WGSL/TypeDeclarations.rb:
* Source/WebGPU/WGSL/generator/main.rb:
* Source/WebGPU/WGSL/tests/valid/overload.wgsl:

Canonical link: <a href="https://commits.webkit.org/262319@main">https://commits.webkit.org/262319@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a96e3e1711483db6cb21b90519bc05c9817315c8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1226 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1260 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1299 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1995 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1096 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1311 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1329 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/1257 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1236 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/1147 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/1141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1862 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1171 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/1130 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/1111 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1121 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/1161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2219 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1169 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1106 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/1140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/304 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1196 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/TypeDeclarations.rb:
```diff
@@ -26,3 +26,32 @@
     # https://bugs.webkit.org/show_bug.cgi?id=254515
 }
+
+operator :vec2, {
```
I don't really know what I'm reading, but OK

litherum commented on Source/WebGPU/WGSL/tests/valid/overload.wgsl:
```diff
@@ -99,3 +99,32 @@ fn testTextureSample() {
   }
 }
+
+fn testVec2() {
```
I still don't quite understand how our testing strategy for WGSL works right now...

tadeuzagallo replied:
These are only validating type checking. All these expressions would fail to type check without this patch. I described it in a different PR, but for now unfortunately this can't be automated, but as soon as I land the standalone compiler and sort a couple other things we can automate these.

tadeuzagallo committed:
[WGSL] Add overloads for vecN constructors
https://bugs.webkit.org/show_bug.cgi?id=254586
rdar://107315663

Reviewed by Myles C. Maxfield.

Add the declarations for all the overloads of vec2, vec3 and vec4 constructors.
This requires introducing a couple new constraints, although constraints aren't
validated yet. There's only one overload per constructor disabled with a fixme,
since they require explicitly specifying the type (e.g. vecN<T>(...)), but we
don't currently support that, but will be implemented next.

* Source/WebGPU/WGSL/Overload.h:
* Source/WebGPU/WGSL/TypeDeclarations.rb:
* Source/WebGPU/WGSL/generator/main.rb:
* Source/WebGPU/WGSL/tests/valid/overload.wgsl:

Canonical link: https://commits.webkit.org/262319@main
